### ISSUE TITLE
Calling `format` with `y` mutates the state

### ISSUE THREAD

**Issue by BroiSatse** [moment/moment#5612]
**Describe the bug**
Calling `format` mutates the state.

**To Reproduce**
1. Go to this fiddle: https://jsfiddle.net/zborwn0u/8/
2. Fiddle logs current time as formatted moment `h:mma y h:mma`.
3. Notice that the time after `y` is always 12:00am
4. Then, it logs `h:mma` again, and `toISOString()` to display 12:00am, proving mutated state.

**Expected behavior**
Calling format should not change the state of moment object.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Browser: Chrome

Please run the following code in your environment and include the output:
```
console.log((new Date()).toString())
console.log((new Date()).toLocaleString())
console.log((new Date()).getTimezoneOffset())
console.log(navigator.userAgent)
console.log(moment.version)
```

```
Sun Jun 21 2020 00:19:01 GMT+0100 (British Summer Time)
21/06/2020, 00:19:01
-60
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.106 Safari/537.36
2.25.3
```

**Comment by NotMoni** [comment]
From my research on this issue. It looks like it has to do with the `y` or year format options.

```js
let m = moment()
let x = moment()

console.log(m.format('y h:mma h:mma y')) // "2020 12:00am 12:00am 2020"
console.log(x.format('h:mma'))           // "8:39pm"
```

but.....

```js
let m = moment()

console.log(m.format('h:mma'))            //  "8:40pm"
console.log(m.format('y h:mma h:mma y'))  // "2020 12:00am 12:00am 2020"
console.log(m.format('h:mma'))            // "12:00am"
```

Hope this helps 🦄 

My environment was the same JSFiddle 

```
"Sat Jun 20 2020 20:43:56 GMT-0400 (Eastern Daylight Time)"
"6/20/2020, 8:43:56 PM"
240
"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.5 Safari/605.1.15"
"2.27.0"
```

**Comment by MinusFour** [comment]
There are a couple of instances on the `era.js` file where they mutate the moment object.

https://github.com/moment/moment/blob/4b615b9d4b00808c16c06c69cfcaa4ecdaedd7b4/src/lib/units/era.js#L167

https://github.com/moment/moment/blob/4b615b9d4b00808c16c06c69cfcaa4ecdaedd7b4/src/lib/units/era.js#L187

https://github.com/moment/moment/blob/4b615b9d4b00808c16c06c69cfcaa4ecdaedd7b4/src/lib/units/era.js#L210

All can be possibly fixed with: 

```js
val = moment(this).startOf('day').valueOf()
```

**Comment by marwahaha** [comment]
I think this is fixed in 2.28.0
